### PR TITLE
Further speed improvements

### PR DESCRIPTION
**Pull request by folkertdev**
and also a bug fix, see comments

### CODE REVIEW

File: src/SHA1.elm [hunk 1]
```diff
@@ -343,27 +338,33 @@ reduceChunkHelp (State initial) b1 b2 b3 b4 b5 b6 b7 b8 b9 b10 b11 b12 b13 b14 b
...
             reduceWords 0 initialDeltaState b1 b2 b3 b4 b5 b6 b7 b8 b9 b10 b11 b12 b13 b14 b15 b16
     in
-    State (Tuple5 (initial.a + a) (initial.b + b) (initial.c + c) (initial.d + d) (initial.e + e))
+    State
+        { a = initial.a + a
+        , b = initial.b + b
+        , c = initial.c + c
+        , d = initial.d + d
+        , e = initial.e + e
+        }
```
Comment (folkertdev) [inline src/SHA1.elm:367]: A plain record is much faster than using the alias (which is a function internally, the record literal is really an object literal in JS)

File: src/SHA1.elm [hunk 2]
```diff
@@ -373,24 +374,84 @@ So in the recursion, `b16` is dropped, all the others shift one position to the
 Then the `deltaState` is also updated with the `value`.
 
 -}
-reduceWords i deltaState b16 b15 b14 b13 b12 b11 b10 b9 b8 b7 b6 b5 b4 b3 b2 b1 =
-    if (i - blockSize) < 0 then
+reduceWords i ((DeltaState { a, b, c, d, e }) as deltaState) b16 b15 b14 b13 b12 b11 b10 b9 b8 b7 b6 b5 b4 b3 b2 b1 =
```
Comment (folkertdev) [inline src/SHA1.elm:377]: This function now performs 8 calculations per iteration.That means fewer function calls, but diminishing returns mean that it's not worth it to go up to 16.

File: src/SHA1.elm [hunk 3]
```diff
@@ -373,24 +374,84 @@ So in the recursion, `b16` is dropped, all the others shift one position to the
...
+                    _ ->
+                        Bitwise.xor b (Bitwise.xor c d) + 0xCA62C1D6
+
+            shiftedA =
+                -- rotateLeftBy 5 a
+                Bitwise.or (Bitwise.shiftRightZfBy (32 - 5) a) (Bitwise.shiftLeftBy 5 a)
+
+            newA =
+                (shiftedA + f + e + int)
+                    |> Bitwise.shiftRightZfBy 0
```
Comment (folkertdev) [inline src/SHA1.elm:445]: This `shiftRightZfBy 0` fixes the bug on the large file reported in #5

File: src/SHA1.elm [hunk 4]
```diff
@@ -419,6 +485,12 @@ rotateLeftBy amount i =
         |> Bitwise.shiftRightZfBy 0
 
 
+rotateLeftBy1 : Int -> Int
+rotateLeftBy1 i =
+    -- because of how `rotateLeftBy1` is used, the `Bitwise.shiftRightZfBy 0` is not required
+    Bitwise.or (Bitwise.shiftRightZfBy 31 i) (Bitwise.shiftLeftBy 1 i)
```
Comment (folkertdev) [inline src/SHA1.elm:491]: specializing to shift by 1 is important. This function takes only 1 argument now, an so elm does not need to apply wrappers that enable currying of this function. Much faster.

**Review by folkertdev — COMMENTED** [review]
On my machine this now does ~16MB/s in a benchmark, and the 90mb file takes just over 5 seconds of scripting time.

**Comment by TSFoster** [comment]
Brilliant, thanks. Love learning about elm perf from you, thanks for all the explanations

**Comment by TSFoster** [comment]
Now published as [2.1.1](https://package.elm-lang.org/packages/TSFoster/elm-sha1/2.1.1/)

(cc @BKSpurgeon)

**Pull request merged**